libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/RustClientSessionDelegate.kt

```diff
@@ -0,0 +1,124 @@
+import org.matrix.rustcomponents.sdk.Session
+import timber.log.Timber
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * This class is responsible for handling the session data for the Rust SDK.
+ *
+ * It implements both [ClientSessionDelegate] and [ClientDelegate] to react to session data updates and auth errors.
+ *
+ * IMPORTANT: you must set the [client] property as soon as possible so [didReceiveAuthError] can work properly.
+ */
+@OptIn(ExperimentalCoroutinesApi::class)
+class RustClientSessionDelegate(
+    private val sessionStore: SessionStore,
+    private val appCoroutineScope: CoroutineScope,
+    coroutineDispatchers: CoroutineDispatchers,
+) : ClientSessionDelegate, ClientDelegate {
+    private val clientLog = Timber.tag("$this")
+
+    // Used to ensure several calls to `didReceiveAuthError` don't trigger multiple logouts
+    private val isLoggingOut = AtomicBoolean(false)
+
+    // To make sure only one coroutine affecting the token persistence can run at a time
+    private val updateTokensDispatcher = coroutineDispatchers.io.limitedParallelism(1)
+
+    /** This [RustMatrixClient] needs to be set up as soon as possible so `didReceiveAuthError` can work properly. */
+    var client: RustMatrixClient? = null
+
+    override fun saveSessionInKeychain(session: Session) {
+        appCoroutineScope.launch(updateTokensDispatcher) {
+            val existingData = sessionStore.getSession(session.userId) ?: return@launch
+            val (anonymizedAccessToken, anonymizedRefreshToken) = session.anonymizedTokens()
+            clientLog.d(
+                "Saving new session data with token: access token '$anonymizedAccessToken' and refresh token '$anonymizedRefreshToken'. " +
+                    "Was token valid: ${existingData.isTokenValid}"
+            )
+            val newData = session.toSessionData(
+                isTokenValid = true,
+                loginType = existingData.loginType,
+                passphrase = existingData.passphrase,
+                sessionPaths = existingData.getSessionPaths(),
+            )
+            sessionStore.updateData(newData)
+            clientLog.d("Saved new session data with access token: '$anonymizedAccessToken'.")
+        }.invokeOnCompletion {
+            if (it != null) {
+                clientLog.e(it, "Failed to save new session data.")
+            }
+        }
+    }
+
+    override fun didReceiveAuthError(isSoftLogout: Boolean) {
+        clientLog.w("didReceiveAuthError(isSoftLogout=$isSoftLogout)")
+        if (isLoggingOut.getAndSet(true).not()) {
+            clientLog.v("didReceiveAuthError -> do the cleanup")
+            // TODO handle isSoftLogout parameter.
+            appCoroutineScope.launch(updateTokensDispatcher) {
+                val currentClient = this@RustClientSessionDelegate.client
+                if (currentClient == null) {
+                    clientLog.w("didReceiveAuthError -> no client, exiting")
+                    return@launch
```

Comment:
Maybe when this happen set `isLoggingOut` back to false, to get another chance to do the cleanup later?